client/core/core.go

```diff
@@ -1647,23 +1648,57 @@ func handleEpochOrderMsg(_ *Core, dc *dexConnection, msg *msgjson.Message) error
 		return fmt.Errorf("epoch order note unmarshal error: %v", err)
 	}
 
-	dc.booksMtx.Lock()
-	defer dc.booksMtx.Unlock()
+	dc.booksMtx.RLock()
+	defer dc.booksMtx.RUnlock()
 
 	ob, ok := dc.books[note.MarketID]
 	if !ok {
 		return fmt.Errorf("no order book found with market id %q",
 			note.MarketID)
 	}
 
+	// Reset an initialized queue if a new order with a different epoch is
+	// received.
+	epoch := ob.Epoch()
+	if epoch != note.Epoch && epoch != 0 {
+		ob.ResetEpoch()
+	}
+
 	return ob.Enqueue(&note)
 }
 
+// handleMatchProofMsg is called when a match_proof notification is received.
+func handleMatchProofMsg(_ *Core, dc *dexConnection, msg *msgjson.Message) error {
+	var note msgjson.MatchProofNote
+	err := msg.Unmarshal(&note)
+	if err != nil {
+		return fmt.Errorf("match proof note unmarshal error: %v", err)
+	}
+
+	dc.booksMtx.RLock()
+	defer dc.booksMtx.RUnlock()
+
+	ob, ok := dc.books[note.MarketID]
+	if !ok {
+		return fmt.Errorf("no order book found with market id %q",
+			note.MarketID)
+	}
+
+	// Reset the epoch queue after processing the match proof message.
+	defer ob.ResetEpoch()
+
+	// TODO: I guess we should be keeping track of the match
+	// proof status for the current epoch refer back to if the
+	// client has an order in the epoch. The client would only
+	// proceed with the match if the match proof is valid. Thoughts?
```

Comment:
I'm not certain of how to handle this. Since the `match_proof` notification is sent in a separate goroutine, it doesn't seem like we can say for certain which will be received first. I would assume the client would wait to act on a match if the match proof has not been validated. The match proof route might then trigger action for waiting matches for which the client is the maker. 

I think the answer will be clear once the matching bits are in place. This is fine for now.